contracts/MuonNodeStaking.sol

```diff
@@ -0,0 +1,624 @@
+            msg.sender
+        );
+        require(node.id != 0, "Node not found for the staker address.");
+
+        require(
+            paidRewardPerToken <= rewardPerToken() ||
+                users[msg.sender].paidRewardPerToken < paidRewardPerToken,
+            "Invalid paidRewardPerToken value."
+        );
+
+        // Verify the authenticity of the withdrawal request.
+        bytes32 hash = keccak256(
+            abi.encodePacked(
+                muonAppId,
+                reqId,
+                msg.sender,
+                users[msg.sender].paidReward,
+                paidRewardPerToken,
+                amount
+            )
+        );
+
+        bool verified = muonVerify(
+            reqId,
+            uint256(hash),
+            signature,
+            muonPublicKey
+        );
+        require(verified, "Invalid signature.");
+
+        if (node.active) {
+            require(amount <= earned(msg.sender), "Invalid withdrawal amount.");
+        } else {
+            require(
+                amount <= users[msg.sender].pendingRewards,
+                "Invalid withdrawal amount."
+            );
+        }
+
+        users[msg.sender].pendingRewards = 0;
+        users[msg.sender].paidReward += amount;
+        users[msg.sender].paidRewardPerToken = paidRewardPerToken;
+        withdrawRequests[reqId] = true;
+        muonToken.transfer(msg.sender, amount);
+        emit RewardGot(reqId, msg.sender, amount);
+    }
+
+    /**
+     * @dev Allows stakers to request to exit from the network.
+     * Stakers can withdraw the staked amount after the exit pending period has passed.
+     */
+    function requestExit() public updateReward(msg.sender) {
+        IMuonNodeManager.Node memory node = nodeManager.stakerAddressInfo(
+            msg.sender
+        );
+        require(node.id != 0, "Node not found for the staker address.");
+
+        require(node.active, "The node is already deactivated.");
+
+        require(
+            users[msg.sender].balance > 0,
+            "No staked balance available for withdrawal."
+        );
+
+        totalStaked -= users[msg.sender].balance;
+        nodeManager.deactiveNode(node.id);
+        emit ExitRequested(msg.sender);
+    }
+
+    /**
+     * @dev Allows stakers to withdraw their staked amount after exiting the network and exit pending period has passed.
+     */
+    function withdraw() public {
+        IMuonNodeManager.Node memory node = nodeManager.stakerAddressInfo(
+            msg.sender
+        );
+        require(node.id != 0, "Node not found for the staker address.");
+
+        require(
+            !node.active &&
+                (node.endTime + exitPendingPeriod) < block.timestamp,
+            "The exit time has not been reached yet."
+        );
+
+        require(
+            !lockedStakes[msg.sender],
+            "Your stake is currently locked and cannot be withdrawn."
+        );
+
+        uint256 amount = users[msg.sender].balance;
```

Comment:
Should it be `balance+pendingRewards`?
or maybe it is better to move `balance+pendingRewards` to `user.withdrawable` on `requestExit`

Reply:
`pendingRewards` is the maximum rewards. It needs a signature from Muon app.